### PR TITLE
feat: no placement in symbols

### LINKED ISSUE #226: Refactor AbstractGate to split placement and gate information
# Overview

We want to split the `AbstractGate` interface into two parts. The placement information and symbol information.

## Example

```julia
struct PlacedGate begin
    gate::AbstractGate
    connected_qubits::Vector{Int}
end

struct H <: AbstractGate begin
end

get_operator(gate::H)::AbstractOperator = DenseOperator([...])
get_num_connected_qubits(gate::H)::Int = 1

struct Rz <: AbstractGate begin
    phase::Real
end
```

## Justification

We currently group both placement information and the gate symbol into the `AbstractGate` interface. The only way to move a qubit given only the interface was to define a `MovedGate`, which embeds the original gate and exposes the interface, but with different connectivity. This seemed to work well. Then @npoirie1 pointed out that the wrapper type interferes with Julia's dispatch rules. We found missing test cases where `MovedGate`s were not tested and caused bugs when passed to some transpilers. To keep with the current architecture, all transpilers (the transpilers in Snowflake and those that are going to be defined by users), would have to handle moved gates correctly. This adds more work every time a new transpiler is written. An alternative is to make `move_gate` part of the `AbstractGate` interface, but then it adds work every time a new gate is written. The root issue was that we coupled the placement information and the gate symbol by making them part of the same interface. The long term solution is to split the interface.

See #224 for more information.

### PR DESCRIPTION
**Pull request by plerouxanyon**
Closes #226

### CODE REVIEW

File: src/core/quantum_gate.jl [hunk 1]
```diff
@@ -359,87 +284,59 @@ Underlying data ComplexF64:
...
-        connected_qubits=vcat(control_qubits,kernel_qubits)
+end
 
-        @assert length(connected_qubits)==
-            length(unique(connected_qubits)) "Control and kernel qubits must all be distinct."*
-            " Received control_qubits: $control_qubits, target_qubits $kernel_qubits"
+controlled(symbol::AbstractGateSymbol, num_control_qubits::Int) = Controlled(symbol, num_control_qubits)
 
+function controlled(gate::Gate, control_qubits::Vector{Int}) 
+    symbol = get_gate_symbol(gate)
```
[the remark this answers is not in the document]
Comment (plerouxanyon) [reply src/core/quantum_gate.jl:306]: Added a check to the constructor of gate and added tests to both `Gate` and `controlled`

File: src/core/quantum_gate.jl [hunk 2]
```diff
@@ -128,18 +126,12 @@ abstract type AbstractGateSymbol end
...
 
-# TODO(#226): delete on completion
-get_connected_qubits(gate::AbstractGateSymbol)=[gate.target]
-
 get_gate_parameters(gate::AbstractGateSymbol)=Dict{String,Real}()
 
-get_num_connected_qubits(gate::AbstractGateSymbol)=length(get_connected_qubits(gate))
-
-get_num_target_qubits(gate::AbstractControlledGateSymbol)=length(get_target_qubits(gate))
+# TODO: Change default to throw not implemented
```
[the remark this answers is not in the document]
Comment (plerouxanyon) [reply src/core/quantum_gate.jl:131]: I've added a follow up issue where we can address it (#293) and added that to the TODO comment

File: src/core/quantum_gate.jl [hunk 3]
```diff
@@ -359,87 +284,59 @@ Underlying data ComplexF64:
...
-    kernel_qubits::Vector{Int}
+struct Controlled{GateType<:AbstractGateSymbol} <: AbstractControlledGateSymbol
+    target::GateType
+    num_control_qubits::Int
 
-    function ControlledGate(
-        kernel::AbstractGateSymbol,
-        control_qubits::Vector{Int},
+    function Controlled(
+        target::AbstractGateSymbol,
```
[the remark this answers is not in the document]
Comment (plerouxanyon) [reply src/core/quantum_gate.jl:292]: Renamed to kernel

File: src/core/transpile.jl [hunk 4]
```diff
@@ -424,11 +424,11 @@ function transpile(::CastSwapToCZGateTranspiler, circuit::QuantumCircuit)::Quant
     qubit_count=get_num_qubits(circuit)
     output=QuantumCircuit(qubit_count=qubit_count)
 
-    for placement in get_circuit_gates(circuit)
-        if is_gate_type(get_gate_symbol(placement), Swap)
-            push!(output, cast_to_cz(get_gate_symbol(placement), get_connected_qubits(placement))...)
+    for gate in get_circuit_gates(circuit)
+        if get_gate_symbol(gate) isa Swap
```
[the remark this answers is not in the document]
Comment (plerouxanyon) [reply src/core/transpile.jl:428]: I updated everywhere I could find where it is applicable

**Comment by plerouxanyon** [comment]
@fSosaRey This branch is ready for review

**Review by fSosaRey — APPROVED** [review]
Good to go!

**Pull request merged**